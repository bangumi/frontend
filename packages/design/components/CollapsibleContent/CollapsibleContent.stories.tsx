import type { ComponentMeta, Story } from '@storybook/react';
import React, { useState } from 'react';

import { render as renderBBCode } from '@bangumi/utils';

import type { CollapsibleContentProps } from '.';
import CollapsibleContent from '.';

const storyMeta: ComponentMeta<typeof CollapsibleContent> = {
  title: 'modern/CollapsibleContent',
  component: CollapsibleContent,
};

export default storyMeta;

const Template: Story<CollapsibleContentProps> = (args) => {
  const [collapsed, setCollapsed] = useState(true);
  return <CollapsibleContent {...args} collapsed={collapsed} onChange={setCollapsed} />;
};

export const Basic = Template.bind({});
const description = `欢迎加入番组WIKI计划小组！

本小组旨在提供和修正 Bangumi 内条目信息的内容，交流 Bangumi 的使用心得，使网站提供的信息更加准确和完善。

如果你对向网友无偿贡献感兴趣并有足够的时间，本小组将对您的修订工作做出指导与规范，同时也欢迎您加入到指导与规范的工作中来。

点击以下链接，查看最近修改的条目，以及有哪些条目有待完善：
[url]http://bangumi.tv/wiki[/url]

参与编辑前需要了解的基本知识：
◆ [url=http://bgm.tv/group/topic/312100]收录范围[/url]
◆ [url=http://bgm.tv/group/topic/24657]成为维基人[/url]
◆ [url=http://bgm.tv/group/topic/15693]相关规则汇总[/url]
◆ [url=http://bgm.tv/group/topic/10657]添加互助[/url]

需要维基管理员参与的：
◆ [url=http://bgm.tv/group/topic/8465]重复条目汇报[/url]
◆ [url=http://bgm.tv/group/topic/8467]错误章节汇报[/url]
◆ [url=http://bgm.tv/group/topic/8466]重复人物汇报[/url]
◆ [url=http://bgm.tv/group/topic/25312]疑似违规条目汇报[/url]

地图：
[url=http://bgm.tv]Bangumi[/url]
├ [url=http://bgm.tv/group/forum]站务论坛[/url]
├ [url=http://bgm.tv/group/wiki]番組WIKI計画[/url]  (就是这里)
└ [url=http://bgm.tv/group/issues]BUG追踪[/url]
`;
Basic.args = {
  threshold: 192,
  content: renderBBCode(description),
};
