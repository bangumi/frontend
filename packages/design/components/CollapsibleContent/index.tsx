import './style/index.less';

import classNames from 'classnames';
import React from 'react';

import { ArrowDownCircle, ArrowUpCircle } from '@bangumi/icons';

import Button from '../Button';

export interface CollapsibleContentProps {
  /** 高度阈值，折叠超过此高度的内容，内容高度小于阈值时禁用折叠 */
  threshold?: number;
  content: React.ReactNode;
  collapsed?: boolean;
  containerClassName?: string;
  onChange?: (collapsed: boolean) => void;
}

const CollapsibleContent: React.FC<CollapsibleContentProps> = ({
  content,
  threshold = 200,
  collapsed = true,
  onChange,
  containerClassName,
}) => {
  const [enableCollapse, setEnableCollapse] = React.useState(false);
  const contentRef = React.useRef<HTMLDivElement>(null);

  React.useLayoutEffect(() => {
    if (!contentRef.current) {
      return;
    }
    setEnableCollapse(contentRef.current.scrollHeight > threshold);
  }, [threshold]);

  const collapsedStyle: React.CSSProperties = {
    maxHeight: threshold,
    overflow: 'hidden',
  };

  const handleUncollapse = (): void => {
    onChange?.(false);
  };

  const handleCollapse = (): void => {
    onChange?.(true);
  };

  const renderControl = (): React.ReactElement | null => {
    if (!enableCollapse) {
      return null;
    }

    if (collapsed) {
      return (
        <div>
          ……
          <Button type='plain' onClick={handleUncollapse}>
            展开
            <ArrowDownCircle />
          </Button>
        </div>
      );
    }
    return (
      <Button type='plain' onClick={handleCollapse}>
        收起
        <ArrowUpCircle />
      </Button>
    );
  };

  return (
    <div className={classNames('bgm-collapsible-content', containerClassName)}>
      <div ref={contentRef} style={enableCollapse && collapsed ? collapsedStyle : undefined}>
        {content}
      </div>
      {renderControl()}
    </div>
  );
};

export default CollapsibleContent;
