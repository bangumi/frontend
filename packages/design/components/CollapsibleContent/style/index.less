.bgm-collapsible-content {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  gap: 0.5rem;
  line-height: 24px;
}
