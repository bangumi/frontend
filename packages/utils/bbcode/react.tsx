import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
