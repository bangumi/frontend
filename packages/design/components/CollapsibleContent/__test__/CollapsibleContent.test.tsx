import { fireEvent, render } from '@testing-library/react';
import React, { useState } from 'react';

import type { CollapsibleContentProps } from '../index';
import CollapsibleContent from '../index';

const Template = (args: CollapsibleContentProps) => {
  const [collapsed, setCollapsed] = useState(args.collapsed);
  return <CollapsibleContent collapsed={collapsed} onChange={setCollapsed} {...args} />;
};

const mockScrollHeight = (height: number) => {
  vi.spyOn(HTMLDivElement.prototype, 'scrollHeight', 'get').mockImplementation(() => height);
};

it('should render', () => {
  const { container } = render(<Template content='bgm' />);
  expect(container.firstChild).toMatchSnapshot();
});

it('should uncollapse', () => {
  mockScrollHeight(10);
  const onChange = vi.fn();
  const { getByText, container } = render(
    <CollapsibleContent content='bgm' threshold={1} collapsed onChange={onChange} />,
  );
  const contentElement = container.firstChild?.firstChild;
  const button = getByText('展开');
  expect(button).toBeInTheDocument();
  expect(contentElement).toHaveStyle({ 'max-height': '1px' });
  fireEvent.click(button);
  expect(onChange).toBeCalledWith(false);
});

it('should collapse', () => {
  mockScrollHeight(10);
  const onChange = vi.fn();
  const { getByText, container } = render(
    <CollapsibleContent content='bgm' threshold={1} collapsed={false} onChange={onChange} />,
  );
  const contentElement = container.firstChild?.firstChild;
  const button = getByText('收起');
  expect(button).toBeInTheDocument();
  expect(contentElement).not.toHaveStyle({
    'max-height': '1px',
  });
  fireEvent.click(button);
  expect(onChange).toBeCalledWith(true);
});
