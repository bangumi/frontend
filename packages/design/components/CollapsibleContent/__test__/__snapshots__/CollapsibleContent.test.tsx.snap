// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render 1`] = `
<div
  class="bgm-collapsible-content"
>
  <div>
    bgm
  </div>
</div>
`;
